test/charms/main/lib/charm.py

```diff
@@ -0,0 +1,165 @@
+    pass
+
+
+class DBRelationDepartedEvent(EventBase):
+    pass
+
+
+class DBRelationBrokenEvent(EventBase):
+    pass
+
+
+class MonRelationJoinedEvent(EventBase):
+    pass
+
+
+class MonRelationChangedEvent(EventBase):
+    pass
+
+
+class MonRelationDepartedEvent(EventBase):
+    pass
+
+
+class MonRelationBrokenEvent(EventBase):
+    pass
+
+
+class HARelationJoinedEvent(EventBase):
+    pass
+
+
+class HARelationChangedEvent(EventBase):
+    pass
+
+
+class HARelationDepartedEvent(EventBase):
+    pass
+
+
+class HARelationBrokenEvent(EventBase):
+    pass
+
+
+class MainCharmEvents(CharmEvents):
+    db_relation_joined = Event(DBRelationJoinedEvent)
+    db_relation_changed = Event(DBRelationChangedEvent)
+    db_relation_departed = Event(DBRelationDepartedEvent)
+    db_relation_broken = Event(DBRelationBrokenEvent)
+    mon_relation_joined = Event(MonRelationJoinedEvent)
+    mon_relation_changed = Event(MonRelationChangedEvent)
+    mon_relation_departed = Event(MonRelationDepartedEvent)
+    mon_relation_broken = Event(MonRelationBrokenEvent)
+    ha_relation_joined = Event(HARelationJoinedEvent)
+    ha_relation_changed = Event(HARelationChangedEvent)
+    ha_relation_departed = Event(HARelationDepartedEvent)
+    ha_relation_broken = Event(HARelationBrokenEvent)
+
+
+class Charm(CharmBase):
+
+    on = MainCharmEvents()
+
+    def __init__(self, framework, key):
+        super().__init__(framework, key)
+
+        # Load attributes that might have been passed via an environment
+        # variable (used for altering charm behavior during testing).
+        charm_config = os.environ.get('CHARM_CONFIG')
+        if charm_config is not None:
+            self._charm_config = pickle.loads(base64.b64decode(charm_config))
+        else:
+            self._charm_config = {}
+
+        self._state_file = self._charm_config.get('STATE_FILE')
+        self._state = {}
+        self._state['event_type_names_observed'] = list()
+
+        # __init__ does not use reset_type because the goal is to
+        # avoid those values being reset on every object instantiation.
+
+        charm_events = dict()
+        charm_events.update({event_name: instance for event_name, instance
+                             in CharmEvents.__dict__.items()
+                             if type(instance) == Event})
+        charm_events.update({event_name: instance for event_name, instance
+                             in type(self.on).__dict__.items()
+                             if type(instance) == Event})
+
+        event_names = charm_events.keys()
+
+        # Set up handlers default handlers for base events.
+        for event_name in event_names:
+            handler_name = f'on_{event_name}'
+            try:
+                getattr(self, handler_name)
+            except AttributeError:
+                setattr(self, handler_name, self._handler_core)
+            self.framework.observe(getattr(self.on, event_name), self)
```

Comment:
Same as above.. feels super introspective, but it's not clear this is necessary. Am I missing something?

Reply:
See the review comment below on event names.